### PR TITLE
Make uint types (un)serializable

### PR DESCRIPTION
**Pull request by shesek**

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
utack

**Comment by elichai** [comment]
We don't use `serde_derive` so sadly you'll have to impl it yourself (I recommend either via one of the macros here: https://github.com/rust-bitcoin/rust-bitcoin/blob/7c47c9a341f5665f89d0c1ab6e958e365bff949b/src/internal_macros.rs or if nothing fits (I don't remember what's in there anymore) using `macro-expand` and then prettifying it)

**Comment by shesek** [comment]
@elichai ah-ah, I see. It was working when I tried because serde_derive is in the dev-dependencies. Will look into changing that to use one of the existing macros.

**Comment by shesek** [comment]
@elichai None of the existing macros seem to fit, I ended up implementing a custom serializer/deserializer by hand at https://github.com/rust-bitcoin/rust-bitcoin/pull/511/commits/0e9672eca82c17ad8038856fdd00bf15135d9d4c.

**Comment by apoelstra** [comment]
Can you add a unit test which parses some fixed json vectors? And covers some error/edge cases?

**Comment by shesek** [comment]
@apoelstra Added some fixed vectors and error cases. Did you have any particular edge cases to test in mind?

**Comment by shesek** [comment]
Is there anything else needed to get this through? :)

**Review by sgeisler — COMMENTED** [review]
What's the motivation for using our internal representation (64bit words) for the serde implementation? This could lead to some weird mixed-endianess in binary serde implementations and doesn't look pretty as json. I think treating these bignum types as byte arrays with a hex representations in human redable serde backends would be best if there are no other compatibility requirements.

Also: travis doesn't seem happy, something failed for rust 1.29.0.

**Comment by shesek** [comment]
@sgeisler No particular reason, I like hex representation better too. I changed to it and fixed compatibility with Rust 1.29. (forced-pushed over the previous implementation, which is available for reference at [`202011-uint-serde-jsonarray`](https://github.com/shesek/rust-bitcoin/tree/202011-uint-serde-jsonarray).)

I wasn't sure whether the new `from_be_slice()` should enforce the length or not. I kept the commit that changed it to return a `Result` separate, let me know what you think and I'll squash it in if it seems acceptable.

An alternative to adding `from_be_slice()` is to change the existing `from_be_bytes()` to take an `AsRef<u8>` which will work for both slices and arrays, but this will lose the type checking for the array length.

**Comment by shesek** [comment]
Travis appears to be stuck due to some unrelated issue, it [works](https://travis-ci.org/github/shesek/rust-bitcoin/jobs/752076368) on my fork.

File: src/util/uint.rs [hunk 1]
```diff
@@ -403,12 +428,60 @@ macro_rules! construct_uint {
...
         }
+
+        #[cfg(feature = "serde")]
+        impl $crate::serde::Serialize for $name {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: $crate::serde::Serializer,
+            {
+                use $crate::hashes::hex::ToHex;
+                serializer.serialize_str(&self.to_be_bytes().to_hex())
```
Comment (sgeisler) [inline src/util/uint.rs:439]: This appears to always serialize as a string, even if the output format is binary. This results in suboptimal encodings. We avoid this in [other areas of the code](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/src/serde_utils.rs#L89) by switching the encoding strategy based on the human readability.

Comment (shesek) [reply src/util/uint.rs]: Agreed, implemented in https://github.com/rust-bitcoin/rust-bitcoin/pull/511/commits/d0eed01b254f0758fa12c5f84de35993d75c6fb2.

Testing this required introducing a new (dev only) dependency on a serde binary serializer (bincode). I wasn't sure if that's desirable or not so I kept it in a separate commit for now, will squash if its acceptable.

**Review by sgeisler — APPROVED** [review]
Looks good now. I don't think a new dev-dependency is much of a problem, especially since it seems to build fine with 1.29.0.

ACK 1be22c3ab355845de360111e32ccbe30efe3db24

**Comment by shesek** [comment]
Okay, great! I rebased on master and squashed/reordered some commits. Let me know if I should just squash them all into one.

**Review by sgeisler — APPROVED** [review]
I don't think the rebase would have been neccesary since @apoelstra prefers merge commits anyway, but I reviewed again, still looks good.

re-ACK 3450ca8f03be6e7f094f79e4a445c0a698a9b720

File: src/util/uint.rs [hunk 2]
```diff
@@ -420,6 +434,13 @@ macro_rules! construct_uint {
 construct_uint!(Uint256, 4);
 construct_uint!(Uint128, 2);
 
+/// Uint error
+#[derive(Debug)]
+pub enum Error {
+    /// Invalid slice length
+    InvalidLength(usize, usize), // (expected, actual)
```
Comment (sanket1729) [inline src/util/uint.rs:500]: nit: should be (actual, expected) as per the above usage.

**Review by sanket1729 — COMMENTED** [review]
tACK 4069c838fcec96858fddf1003db5910d603190bc. 

Left a minor nit

File: src/util/uint.rs (hunk 2 above)
Comment (sgeisler) [reply src/util/uint.rs:500]: This should probably also be in the actual doc comment (and of course correct).

**Comment by sgeisler** [comment]
This needs a rebase after #533 anyway, would be great if @sanket1729's nit could be addressed at the same time. I'll re-ack then.

File: src/util/uint.rs (hunk 2 above)
Comment (shesek) [reply src/util/uint.rs]: Good catch, fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/511/commits/46acd7787fbdf135a2c7389f0628f883d7a048da.

**Comment by shesek** [comment]
@sgeisler Rebased and fixed the nit.

**Review by sgeisler — APPROVED** [review]
ACK 46acd7787fbdf135a2c7389f0628f883d7a048da

File: src/util/uint.rs [hunk 3]
```diff
@@ -403,12 +428,78 @@ macro_rules! construct_uint {
...
+            }
+        }
     );
 }
 
 construct_uint!(Uint256, 4);
 construct_uint!(Uint128, 2);
 
+/// Uint error
+#[derive(Debug)]
```
Comment (dr-orlovsky) [inline src/util/uint.rs:497]: I think we can be opportunistic here, as Rust API guidelines suggest:

```suggestion
#[derive(Debug, PartialEq, PartialOrd, Clone, Copy, Hash)]
```

File: src/util/uint.rs [hunk 4]
```diff
@@ -403,12 +428,78 @@ macro_rules! construct_uint {
...
+pub enum Error {
+    /// Invalid slice length (actual, expected)
+    InvalidLength(usize, usize),
+}
```
Comment (dr-orlovsky) [inline src/util/uint.rs:501]: This struct looks wierd to me. Why not

```suggestion
/// Invalid slice length
pub struct ParseLengthError {
    /// The length of the slice de-facto
    pub actual: usize,
    /// The required length of the slice
    pub expected: usize,
}
```

File: src/util/uint.rs [hunk 5]
```diff
@@ -403,12 +428,78 @@ macro_rules! construct_uint {
...
+
```
Comment (dr-orlovsky) [inline src/util/uint.rs:502]: It's a frequent problem for upstream dependencies when a error type returned by some function does not implement `Display` and `Error` traits. Can you pls add them?

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
Sorry for being late, but I see space for further improvements over a type system here. Had _a lot_ of issues with my own code when rust-bitcoin does not implement errors properly.

File: src/util/uint.rs (hunk 3 above)
Comment (sgeisler) [reply src/util/uint.rs:497]: Good point, I always forget theses derives myself :see_no_evil: do you know if there is some way to warn about missing (standard) derives?

File: src/util/uint.rs (hunk 4 above)
Comment (sgeisler) [reply src/util/uint.rs:501]: I guess that would actually make sense here. Even though I also always use `enum`s for errors in case another condition appears. But that seems beyond unlikely here.

File: src/util/uint.rs (hunk 3 above)
Comment (dr-orlovsky) [reply src/util/uint.rs:497]: Even if there is a way, I never heard of it

Comment (dr-orlovsky) [reply src/util/uint.rs:497]: Oh no, maybe clippy may help?

**Comment by shesek** [comment]
I switched to a single-variant error type and implemented the suggested traits, apart from `Error` which I noticed isn't implemented anywhere else in rust-bitcoin. Should I implement that too?

File: src/util/uint.rs [hunk 6]
```diff
@@ -403,12 +428,87 @@ macro_rules! construct_uint {
...
+            }
+        }
     );
 }
 
 construct_uint!(Uint256, 4);
 construct_uint!(Uint128, 2);
 
+/// Invalid slice length
+#[derive(Debug, PartialEq, PartialOrd, Clone, Copy, Hash)]
```
Comment (dr-orlovsky) [inline src/util/uint.rs:497]: As just discussed with @apoelstra in https://github.com/rust-bitcoin/rust-bitcoin/pull/521#discussion_r557669163 we need `Eq` everywhere we have `PartialEq`. I also assume we need `Ord` once we have `PartialOrd`

File: src/util/uint.rs (hunk 5 above)
Comment (dr-orlovsky) [reply src/util/uint.rs:511]:
```suggestion
impl ::std::error::Error for ParseLengthError {}
```

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]

**Comment by dr-orlovsky** [comment]
Not sure about `Error` though would it compile on 1.29. I assume yes, and the reason it is absent everywhere is that it was not present before we upgraded to 1.29

**Comment by thomaseizinger** [comment]
> apart from `Error` which I noticed isn't implemented anywhere else in rust-bitcoin

That doesn't seem to be true? GitHub search returns 12 results for "impl error::Error": https://github.com/rust-bitcoin/rust-bitcoin/search?q=%22impl+error%3A%3AError%22

Please implement std::error::Error otherwise working with libraries like `anyhow` is quite painful.

**Comment by shesek** [comment]
@thomaseizinger Oops, you're of course correct, not sure how my search missed that. I will implement this and the other suggested changes.

File: src/util/uint.rs (hunk 6 above)
Comment (shesek) [reply src/util/uint.rs]: I added `Eq`, but wasn't sure about `Ord`. Should I add it too? We do already derive `PartialOrd`.

**Comment by shesek** [comment]
I implemented `Error` and `Eq`.

**Review by dr-orlovsky — CHANGES_REQUESTED** [review]
I still believe we need `Ord` when we have `Eq` and `PartialOrd`. The rest looks fine for me now

**Comment by sgeisler** [comment]
@dr-orlovsky I don't think that's the case _mathematically speaking_ (it probably still makes sense practically in this case if it's possible).

Let's imagine the set of all binary strings. `Eq` is implemented as expected: compare bit values (and implicitly legth). `PartialOrd` is implemented that it returns `Some(Equal)` if `a == b`, `Some(Greater)` if `a` is a prefix of `b`, `Some(Less)` if `b` is a prefix of `a` and `None` otherwise. This is a valid impl as it's both asymmetric and transitive, compatible with the `Eq` impl, but not `Ord`.

That's obviously a contrived example, but I don't see a strict requirement for `Ord` just because we have `Eq` and `PartialOrd`.

**Comment by sanket1729** [comment]
On one side, if we can derive something for free, we should derive it. 

On the other side, `Display`, `Clone`, `Debug`, `Error`, and `Eq`(implying `PartialEq`) are the only traits that I can foresee being used. 

I don't think it's important what we end up doing, but we should decide one thing and be consistent with that for all the Error types in the repo. This would help when we wrap `Error` of one type as an enum variant of another Enum and try to auto derive on that.

**Comment by dr-orlovsky** [comment]
The use for `Ord` are the situations where you need to collect errors from multiple attempts of processing into a `HashMap`, mapping them to the item. Thus, if we can derive `Ord` I think we should do it, and not only here, but on all `Error` types - which is a matter of other PRs of course (I did one already https://github.com/rust-bitcoin/rust-bitcoin/pull/551)

**Comment by sanket1729** [comment]
I think `HashMap` should only `Hash` and `Eq`, but `BTreeMap` will require `Ord`. It may be possible that we have extra work for manually impl `Ord` because the error contains f64 for example. 

That being said, we can try everything and see how it goes.

**Comment by dr-orlovsky** [comment]
I do agree with @apoelstra in https://github.com/rust-bitcoin/rust-bitcoin/issues/555#issuecomment-764958577 that we should implement both `Ord` (and `PartialOrd` of course) where is possible - and this is also Rust API Guidelines recommendation.

Had plenty of pain not being able to use datatypes in `BTree*` when I need to put them there.

**Comment by shesek** [comment]
I added `Ord`. I see that `Hash` and `Default` were also mentioned in #555, are they relevant for error types?

**Review by dr-orlovsky — APPROVED** [review]
Thank you and sorry for being insistent. utACK 55657cbffb353bd1e70ff2501d1141675529fd7b

**Comment by dr-orlovsky** [comment]
I think here `Default` just makes not sense: you cen't have a "default" Error. And the type already has `Hash` derived on it

**Review by sanket1729 — APPROVED** [review]
tACk a1e98a67964be214edd3e6bf27dbd2820e0a7f19

**Review by sgeisler — APPROVED** [review]
utACK a1e98a67964be214edd3e6bf27dbd2820e0a7f19

@sanket1729 are you still against merging without a more thorough test than CI (=currently @apoelstra's test suite)? I'm currently building my own setup (cargo-all-features and some bash, not as fancy but I'll understand it xD). So I hope we can get stuff merged again soon.

**Comment by sanket1729** [comment]
Yeah, let's merge this.
